File: DSweek2/Itm.java
public class Itm implements Comparator<Student>{
    @Override
    public int compare(Student s1,Student s2){
        if(s1.year!=s2.year){
            return s1.year-s2.year;
        }
        return s1.name.compareTo(s2.name);
    }
    public int countNames(List<Student< students){
        Set<String> names=new HashSet<>();  //HashSet<String< names=new HashSet<>();
        for(Student student:students){
            names.add(student.getname);
        }
        //for(int j=0;j<students.size();j++){
        //    names.add(students.get(j).name);   //students.get(j).toSring();
        return names.size();
    }
    public boolean isBracketBalanced(String s){ //A. for switch B.
        Stack<Character> stack=new Stack<>();
        for(int i=0;i<s.length();i++){
            char c=s.charAt(i);
            if(c=='(' || c=='{' || c=='['){
                stack.push(c);
            }else if(c==')' || c=='}' || c==']'){.   //if
                if(stack.isEmpty()){
                    return false;
                }
                char top=stack.pop();//
                if((c==')' && top!='(') || (c=='}' && top!='{') || (c==']' && top!='[')){
                    return false;
                }
            }
        }
        return stack.isEmpty(); //return true;
    }
}

File: DSweek2_c++/Itm.cpp
// Itm.cpp
// C++ version of Itm class with Student comparator, countNames, and isBracketBalanced
#include <iostream>
#include <vector>
#include <set>
#include <stack>
#include <string>
#include <algorithm>
using namespace std;

class Student {
public:
    int year;
    string name;
    Student(int year, const string& name) : year(year), name(name) {}
};

struct Itm {
    // Comparator for Student
    bool operator()(const Student& s1, const Student& s2) const {
        if (s1.year != s2.year) return s1.year < s2.year;
        return s1.name < s2.name;
    }
    // Count unique names
    int countNames(const vector<Student>& students) {
        set<string> names;
        for (const auto& student : students) {
            names.insert(student.name);
        }
        return names.size();
    }
    // Check if brackets are balanced
    bool isBracketBalanced(const string& s) {
        stack<char> st;
        for (char c : s) {
            if (c == '(' || c == '{' || c == '[') {
                st.push(c);
            } else if (c == ')' || c == '}' || c == ']') {
                if (st.empty()) return false;
                char top = st.top(); st.pop();
                if ((c == ')' && top != '(') || (c == '}' && top != '{') || (c == ']' && top != '['))
                    return false;
            }
        }
        return st.empty();
    }
};

int main() {
    vector<Student> students = { {2025, "Alice"}, {2024, "Bob"}, {2025, "Alice"} };
    Itm itm;
    cout << "Unique names: " << itm.countNames(students) << endl;
    cout << "Balanced ({}[]): " << itm.isBracketBalanced("({[]})") << endl;
    cout << "Balanced ({}[)): " << itm.isBracketBalanced("({[)})") << endl;
    return 0;
}
